### PR TITLE
Add scan, sensor aggregate; add get memory handling

### PR DESCRIPTION
**Pull request by nichtsfrei**
**What**:

Added aggregate handler for registering sensors and creating scans.

To verify it run:

```
docker-compose up --build
```

and watch for registration messages e.g.:

```

director_1    | 2021/06/30 10:00:23 [greenbone.sensor] Publishing message: {"id":"bla","created":533071875,"message_type":"modified.sensor","message_id":"da1d09b4-cbca-452a-b38f-0cacaefa53f8","group_id":"084a9a8e-facc-48b3-a60d-30bf82405369"} 

```

**Why**:

**How**:

**Checklist**:

- [ ] Tests
- [ ] CHANGELOG.md entry
- [ ] Documentation

### CODE REVIEW

File: director/handler/scan/scan.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	} else if scan == nil {
+		log.Printf("Scan %s not found, creating a new one.", m.ID)
+		scan = &models.Scan{
+			ID: m.ID,
+		}
+	}
+	applyTargetID := func(k string, v interface{}) error {
+		switch k {
+		case "target_id":
+			if str, ok := v.(string); ok {
```
Comment (paulnaumann) [inline director/handler/scan/scan.go:95]: extract the logic of this case into a separate method

File: messages/handler/handler.go [hunk 2]
```diff
@@ -37,88 +87,146 @@ type Getter interface {
...
 
-func (mh onAggregate) On(message []byte) (interface{}, error) {
+func asResponse(t string, d interface{}) *connection.SendResponse {
+	return &connection.SendResponse{
+		Topic: t,
+		MSG:   d,
+	}
+}
+
+func (mh onMessage) On(topic string, message []byte) (*connection.SendResponse, error) {
```
Comment (paulnaumann) [inline messages/handler/handler.go:119]: I personally think this method is to long, and should be refactored into multiple smaller ones.

File: models/scan.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
+package models
+
+import "github.com/greenbone/eulabeia/messages"
+
+// Scan contains Target as well as volatile information for a specific scan
+type Scan struct {
+	Target
+	ID      string   `json:"id"`      // ID of a Scan
+	Exclude []string `json:"exclude"` // Exclude hosts from scan
```
Comment (paulnaumann) [inline models/scan.go:9]: why is excluded part of scan, and not part of target?

**Review by paulnaumann — COMMENTED** [review]
Some general ideas, i could exactly match exactly to any code:
- i personally don't like the project structure. code, entrypoints and documentation are stating in the same directory depth. I would prefer it, if at least the code would be moved down into a package like 'src' 
- I don't like the multiple struct definitions in one file
- I shortly checked if sentry will work with your current logging. From what I could find online it seems that way. But this is some thing you should check before adding to much logging.

File: cmd/eulabia-director/main.go [hunk 4]
```diff
@@ -10,32 +10,34 @@ import (
...
 
 func main() {
 	server := flag.String("server", "localhost:1883", "A clientid for the connection")
 	clientid := flag.String("clientid", "", "A clientid for the connection")
 	flag.Parse()
 
-	log.Println("Starting sensor")
-	c, err := mqtt.New(*server, *clientid, "", "")
+	log.Println("Starting director")
+	c, err := mqtt.New(*server, *clientid, "", "", nil)
```
[the remark this answers is not in the document]
Comment (nichtsfrei) [reply cmd/eulabia-director/main.go:26]: >     * i personally don't like the project structure. code, entrypoints and documentation are stating in the same directory depth. I would prefer it, if at least the code would be moved down into a package like 'src'
It is untypical for go to use a src dir, so although you personally don't like it I would rather follow the standard layout of the language.

See: https://github.com/kubernetes/kubernetes for a larger reference.

**Comment by nichtsfrei** [comment]
>     * before adding
>     * i personally don't like the project structure. code, entrypoints and documentation are stating in the same directory depth. I 



> Some general ideas, i could exactly match exactly to any code:
> 
>     * i personally don't like the project structure. code, entrypoints and documentation are stating in the same directory depth. I would prefer it, if at least the code would be moved down into a package like 'src'

It is untypical for go to use a src dir, so although you personally don't like it I would rather follow the standard layout of the language.

See: https://github.com/kubernetes/kubernetes for a larger reference.

> 
>     * I don't like the multiple struct definitions in one file

Ok, I dislike just one struct files because it makes it harder to grep the code especially when they are kind of connected.

> 
>     * I shortly checked if sentry will work with your current logging. From what I could find online it seems that way. But this is some thing you should check before adding to much logging.

I don't understand what you mean.

File: cmd/eulabia-sensor/main.go [hunk 5]
```diff
@@ -1,9 +1,63 @@
...
+	"github.com/greenbone/eulabeia/connection/mqtt"
+	"github.com/greenbone/eulabeia/messages"
+	"github.com/greenbone/eulabeia/messages/handler"
+	"github.com/greenbone/eulabeia/sensor/handler/memory"
 )
 
 func main() {
-	print(eulabeia.Hello())
+	topic := "greenbone.sensor"
+	server := flag.String("server", "localhost:1883", "A clientid for the connection")
```
[the remark this answers is not in the document]
Comment (nichtsfrei) [reply cmd/eulabia-sensor/main.go:21]: There will most likely be a config files, environment variable mix. But unless we have the need we don't implement it. So Although it will most likely change currently we choose the easiest option.

File: cmd/eulabia-director/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nichtsfrei) [reply cmd/eulabia-director/main.go:26]: In this case it is single use variable which is directly used within the main.go and not passed through. But I can accept that :)

File: cmd/eulabia-sensor/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nichtsfrei) [reply cmd/eulabia-sensor/main.go:21]: We will most likely endup with a mix of env variables and configuration file; but currently we don't need it because we have little to no configuration. We will do that when we start with openvas integration since then the structure of the configuration file will be more clear.

File: connection/mqtt/mqtt.go [hunk 6]
```diff
@@ -21,31 +21,42 @@ func (m MQTT) Close() error {
...
-			message, err := h.On(p.Payload)
+func (m MQTT) Register(topic string, handler connection.OnMessage) error {
+
+	m.client.Router.RegisterHandler(topic, func(p *paho.Publish) {
+		log.Printf("[%s] retrieved message: %s", topic, string(p.Payload))
+		message, err := handler.On(topic, p.Payload)
+		if err != nil {
+			panic(err)
+		}
+		if message != nil {
```
[the remark this answers is not in the document]
Comment (nichtsfrei) [reply connection/mqtt/mqtt.go:32]: err is an indicator that there is an issue which must be handled within the recipient itself and message is an optional return message; both are used to verify if the iteration of handler can be stopped or should continue. There is an optimization possibility by extending the return values if the message got handled but does not return something to stop the iterarion; but so far we don't have an issue with the amount of handler so I considered it premature optimization.

File: director/handler/scan/scan.go (hunk 1 above)
Comment (nichtsfrei) [reply director/handler/scan/scan.go:95]: I would leave it as closure because it is using scan of the upper context directly and does not follow the same contract as Modify that would mean either:
- extend the struct with scan pointer
- build a function to return a closure
I think both options are harder to read than just using a closure directly.

**Comment by paulnaumann** [comment]
> > * I shortly checked if sentry will work with your current logging. From what I could find online it seems that way. But this is some thing you should check before adding to much logging.
> > ```
> 
> I don't understand what you mean.

What I was trying to say was, that it might be good to check if sentry works with the default log calls, you are using right now. 
Because if it does not, you might need to replace more lines later.

File: director/handler/scan/scan.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+// Storage is for poutting and getting a models.Scan
+type Storage interface {
+	Put(models.Scan) error            // Overrides existing or creates a models.Scan
+	Get(string) (*models.Scan, error) // Gets a models.Scan via ID
+	Delete(string) error
+}
+
+// Depositary stores models.Scan as json within a given StorageDir
+// The filename is a uuid without suffix.
+type Depositary struct {
```
[the remark this answers is not in the document]
Comment (nichtsfrei) [reply director/handler/scan/scan.go:26]: Probably more about style; I rather see definitions at one glance especially if they're used in the same context and don't have meaning outside of that.

File: models/scan.go (hunk 3 above)
Comment (nichtsfrei) [reply models/scan.go:9]: Actually we need both; Scan for start and stop and within Target to exclude hosts from a given range. That will be changed in scan.finished and target.exclude; initially I thought exclude was just used for start, stop, resume hence on scan.

**Comment by nichtsfrei** [comment]
I thought we are using sentry just for errors and not as an log aggregate? The overall approach we are following currently is passing errors through to the caller when it cannot be handled so the integration should just be done within the actual programs; if there's a panic we also need to recover it as late as possible meaning in the program itself.

File: cmd/eulabia-sensor/main.go [hunk 8]
```diff
@@ -1,9 +1,63 @@
...
+	}
+	c.Publish(topic, messages.Modify{
+		Message: messages.NewMessage("modify.sensor", "", ""),
+		ID:      *sensorID,
+		Values: map[string]interface{}{
+			"type": "undefined",
+		},
+	})
+	mh := handler.New(handler.FromGetter(memory.New()))
+	err = c.Subscribe(map[string]connection.OnMessage{topic: mh})
```
Comment (Kraemii) [inline cmd/eulabia-sensor/main.go:49]: More comments about what is actually happening. It is hard to read functions which are nested 3 times into each other.

**Comment by Kraemii** [comment]
The overall structure seems to be a bit complicated as we have many modules just named handler.go. Therefore it is hard to follow what is actually used and which module is responsible for what.

**Comment by nichtsfrei** [comment]
> The overall structure seems to be a bit complicated as we have many modules just named handler.go. Therefore it is hard to follow what is actually used and which module is responsible for what.

Agree, do you have a proposal? For `messages/handler` I think it's fine because it is meant as an abstraction. Otherwise I run out of ideas because they're essentially specific implementation of message handler for a specific kind e.g. sensor, director. 

I could separate them between `aggregate` and `events` but that seems a bit artificial and it wouldn't solve the overall repetition issue.

File: messages/handler/handler.go (hunk 2 above)
Comment (nichtsfrei) [reply messages/handler/handler.go]: Refactored with: a384f09e69e4a083f72fef3eeee4390ab521ed5c

File: cmd/eulabia-sensor/main.go (hunk 8 above)
Comment (nichtsfrei) [reply cmd/eulabia-sensor/main.go]: Removed the need for nesting.

File: cmd/eulabia-sensor/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nichtsfrei) [reply cmd/eulabia-sensor/main.go:21]: There is an own task for it; hence I close it.

File: director/handler/scan/scan.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (nichtsfrei) [reply director/handler/scan/scan.go]: We had a discussion about it; since we are jumping between go and c and it is not that uncommon for go to have multiple structs defined we decided that we keep it for now but will further separate when:
- it's getting out of hand
- structs don't share a common domain

So although it will get redefined most likely not within this pr.

File: models/scan.go (hunk 3 above)
Comment (nichtsfrei) [reply models/scan.go]: Resolved by: f394486000c84714fe14693462a9b33593109dda

File: cmd/eulabia-director/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nichtsfrei) [reply cmd/eulabia-director/main.go]: Renamed to client 0d2d820eda2d3cec7b55d0b2a2ab890065fafc71 although there are various variants left I would like to try to follow that new rule as closely as possible and refactor is step by step to speed up the review process as a whole going to resolve this conversation.

**Review by Kraemii — APPROVED** [review]
